iroh-gossip/src/proto/plumtree.rs

```diff
@@ -0,0 +1,603 @@
+        for ihave in ihaves {
+            if !self.received_messages.contains(&ihave.id) {
+                self.missing_messages
+                    .entry(ihave.id)
+                    .or_default()
+                    .push_back((sender, ihave.round));
+
+                if !self.graft_timer_scheduled.contains(&ihave.id) {
+                    self.graft_timer_scheduled.insert(ihave.id);
+                    io.push(OutEvent::ScheduleTimer(
+                        self.config.graft_timeout_1,
+                        Timer::SendGraft(ihave.id),
+                    ));
+                }
+            }
+        }
+    }
+
+    fn on_send_graft_timer(&mut self, id: MessageId, io: &mut impl IO<PA>) {
+        if self.received_messages.contains(&id) {
+            return;
+        }
+        let entry = self
+            .missing_messages
+            .get_mut(&id)
+            .and_then(|entries| entries.pop_front());
+        if let Some((peer, round)) = entry {
+            self.add_eager(peer);
+            let message = Message::Graft(Graft {
+                id: Some(id),
+                round,
+            });
+            io.push(OutEvent::SendMessage(peer, message));
+
+            // "when a GRAFT message is sent, another timer is started to expire after a certain timeout,
+            // to ensure that the message will be requested to another neighbor if it is not received
+            // meanwhile. This second timeout value should be smaller that the first, in the order of
+            // an average round trip time to a neighbor." (p9)
+            io.push(OutEvent::ScheduleTimer(
+                self.config.graft_timeout_2,
+                Timer::SendGraft(id),
+            ));
+        }
+    }
+
+    fn on_graft(&mut self, sender: PA, details: Graft, io: &mut impl IO<PA>) {
+        self.add_eager(sender);
+        if let Some(id) = details.id {
+            if let Some(message) = self.cache.get(&id) {
+                io.push(OutEvent::SendMessage(
+                    sender,
+                    Message::Gossip(message.clone()),
+                ));
+            }
+        }
+    }
+
+    fn on_neighbor_up(&mut self, peer: PA) {
+        self.add_eager(peer);
+    }
+
+    // "When a neighbor is detected to leave the overlay, it is simple removed from the membership.
+    // Furthermore, the record of IHAVE messages sent from failed members is deleted from the missing history" (p9)
+    fn on_neighbor_down(&mut self, peer: PA) {
+        self.missing_messages.retain(|_message_id, ihaves| {
+            ihaves.retain(|(ihave_peer, _round)| *ihave_peer != peer);
+            !ihaves.is_empty()
+        });
+        self.eager_push_peers.remove(&peer);
+        self.lazy_push_peers.remove(&peer);
+    }
+
+    /// Moves peer into eager set.
+    fn add_eager(&mut self, peer: PA) {
+        self.eager_push_peers.insert(peer);
+        self.lazy_push_peers.remove(&peer);
```

Comment:
I assume this is doing an implicit copy that be prevented by swapping these two lines